### DISCUSSION TITLE
Trouble compiling simple FLTK project

### DISCUSSION THREAD

**Discussion by mwitz8** [fltk-rs/fltk-rs#680] [Q&A] [answered]
0

I'm trying to build and run an example from FLTK's documentation; 
```
use fltk::{app::, button::, frame::, window::};

fn main() {
    let app = App::default();
    let mut wind = Window::new(100, 100, 400, 300, "Hello, FLTK!");
    let mut frame = Frame::new(0, 0, 400, 200, "Boring label");
    let mut but = Button::new(160, 210, 80, 40, "Click me!");

    wind.end();
    wind.show();

    // Remember: Callbacks after initializing the interface
    but.set_callback(move || frame.set_label("Hello, World!"));

    app.run().unwrap();
}
```
But I get the following error:

```
   Compiling fltk-sys v1.0.11
error: failed to run custom build command for `fltk-sys v1.0.11`         

Caused by:
  process didn't exit successfully: `C:\Users\mattw\Documents\engine\target\debug\build\fltk-sys-598a5e0afee061a3\build-script-main` (exit code: 101)
  --- stdout
  cargo:rerun-if-changed=build/android.rs
  cargo:rerun-if-changed=build/bundled.rs
  cargo:rerun-if-changed=build/link.rs
  cargo:rerun-if-changed=build/main.rs
  cargo:rerun-if-changed=build/source.rs
  cargo:rerun-if-changed=build/utils.rs

  --- stderr
  thread 'main' panicked at 'CMake is needed to build the fltk source files!
  Did you intend to use the fltk-bundled feature?', C:\Users\mattw\.cargo\registry\src\github.com-1ecc6299db9ec823\fltk-sys-1.0.11\build\source.rs:8:9
  note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```
I tried installing CMake, but it doesn't seem to help. Could you guys help me figure out where I'm going wrong?

**Comment by MoAlyousef** [comment, marked as answer]
How did you install CMake?
Is it invokable from your shell/terminal?
What's the output of 
```
cmake --version
```

If you're running x64 bit windows, you can try modifying your Cargo.toml
```
[dependencies]
fltk = { version = "1", features = ["fltk-bundled"] }
```
This will use an already prebuilt version of fltk.

**Reply by mwitz8** [reply]
The output tells me I have CMake version 3.20.2 . I'm just using the prebuilt version of fltk as you provided, and that seems to be working. Thanks a ton.

**Reply by kkocdko** [reply]
Hi, which toolchain are you using for "fltk-bundled"? I tried "fltk-bundled" but got error "... undefined reference to `__imp___acrt_iob_func' ...". I google it, seems that this problem is because you use msys-mingw, then rust's buildin mingw version is too low. Maybe you can specify it in the document?

**Reply by MoAlyousef** [reply]
Hi
I use gcc 10 with gnu ld 2.36. Tried on several pc's with msys2 installed and varying versions of toolchains and windows versions. What versions do you have on your system. AFAIK Rust's linking issues on msys2 had been resolved several versions ago.
https://github.com/rust-lang/rust/issues/47048

**Reply by kkocdko** [reply]
`rustc 1.52.1 (9bc8c42bb 2021-05-09) with target x86_64-pc-windows-gnu`
When I just run cargo build, I got that error.
Now I remove rust-mingw and then use tdm-gcc(gcc 9.2), everything well.

**Reply by kkocdko** [reply]
The Rust's linking issues seems only half solved by auto detect the new gcc.exe and ld.exe on %PATH%, but buildin mingw still stay on low version...
